pkg/rbd/nodeserver.go

```diff
@@ -248,21 +296,118 @@ func (ns *NodeServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
 		return nil, status.Error(codes.NotFound, err.Error())
 	}
 	if notMnt {
-		// TODO should consider deleting path instead of returning error,
-		// once all codes become ready for csi 1.0.
-		return nil, status.Error(codes.NotFound, "volume not mounted")
+		if err = os.RemoveAll(targetPath); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		return &csi.NodeUnpublishVolumeResponse{}, nil
+	}
+
+	// TODO remove this once  we stop supporting legacy volume delete/mount
+	isLegacyVolume := false
+	_, err = getVolumeName(req.GetVolumeId())
+	if err != nil {
+		// error ErrInvalidVolID may mean this is an 1.0.0 version volume, check for name
+		// pattern match in addition to error to ensure this is a likely v1.0.0 volume
+		if _, ok := err.(ErrInvalidVolID); !ok || !isLegacyVolumeID(req.GetVolumeId()) {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+
+		_, err = getLegacyVolumeName(req.GetTargetPath())
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+		isLegacyVolume = true
+	}
+	if isLegacyVolume {
+
+		// Unmount the volume
+		devicePath, cnt, devErr := mount.GetDeviceNameFromMount(ns.mounter, targetPath)
+		if devErr != nil {
+			return nil, status.Error(codes.Internal, devErr.Error())
+		}
+
+		// skip if the device not found, during remount the device is mounted to
+		// stageTargetPath
+		if devicePath == "" {
+			klog.Infof("rbd: device not mounted to targetPath %s", targetPath)
+			// Remove targetPath
+			if err = os.RemoveAll(targetPath); err != nil {
+				klog.Errorf("failed to remove targetPath: %s with error: %v", targetPath, err)
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+			return &csi.NodeUnpublishVolumeResponse{}, nil
+		}
+		if err = ns.unmount(targetPath, devicePath, cnt); err != nil {
+			return nil, err
+		}
+		klog.Infof("rbd: successfully unbinded volume %s from %s", req.GetVolumeId(), targetPath)
+
+		return &csi.NodeUnpublishVolumeResponse{}, nil
+	}
+
+	if _, err = execCommand("umount", []string{targetPath}); err != nil {
```

Comment:
Shouldn't this use `ns.mounter.Unmount(targetPath)` instead of calling umount directly?

Reply:
Apologies I think this should be `mount.Unmount`. I was looking at inverse of what we use in the Stage/Publish in Unstage/Unpublish, not sure if Unstage earlier was not using ns.mounter.Unmount (as Stage is using ns.mounter.Mount/FormatAndMount), but here the inverse is `mount.Unmount`